### PR TITLE
[JENKINS-72284] Take agents offline and online immediately

### PR DESCRIPTION
**Pull request by mawinter69**
## [JENKINS-72284] Take agents offline and online immediately

[JENKINS-72284] when an agent with the wrong java version connects, it will be taken offline right away. Previously, the agent was only taken offline when one visited the /computer page.

Use monitor specific offline causes that inherit from `MonitorOfflineCause`. This allows the agent to be brought online again automatically, after it has been restarted with the correct java/remoting version.   [Jenkins 2.433 and later](https://github.com/jenkinsci/jenkins/pull/8618) will then reset the Admin monitor if that was the only reason that the agent was taken offline.

[Jenkins core 2.434 or newer](https://github.com/jenkinsci/jenkins/pull/8593) made 2 methods public that are required for this pull request.

Requires Jenkins 2.440 or newer because that has been selected as the next LTS baseline.

### Testing done

I compared [2.426.3 with the existing release](https://github.com/MarkEWaite/docker-lfs/blob/7ee53b04fff50c6509f606dd87954e857301ae59/plugins.txt#L149), [2.444 with the existing release](https://github.com/MarkEWaite/docker-lfs/blob/8f3ff63450c89eb91f1183647b45955a66453ff9/plugins.txt#L149), and [2.444 with the incremental build](https://github.com/MarkEWaite/docker-lfs/blob/0457286b8cb29cf92cf16e1311f53e86272786e9/plugins.txt#L149) of this plugin.

#### 2.426.3 and current plugin release

* Confirmed that a Pipeline job that requires FreeBSD will use Java 17 on my Jenkins 2.426.3 controller that runs Java 21 (since Java 21 is not yet available as a FreeBSD package):
```groovy
pipeline {
    agent {
        label 'FreeBSD'
    }
    stages {
        stage('FreeBSD-Agent') {
            steps {
                sh 'hostname && uname -a && java -version'
            }
        }
    }
}
```
* Confirmed that when I click the "Apply" button on the "Node monitoring configuration" page on Jenkins 2.426.3 after enabling "Disconnect agent when incompatibility is found", the FreeBSD agents remain connected even though they should have been disconnected.  When the Pipeline job runs, it reports that it was run on a FreeBSD agent and used Java 17
* Confirmed that when I visit the "Computer" page http://mark-pc2.markwaite.net:8080/computer/ then the FreeBSD agents are disconnected and the Pipeline job blocks waiting for a FreeBSD agent
* Confirmed that when I disable the "Disconnect agent when incompatibility is found" setting and click the "Apply" button, the FreeBSD agents remain offline even though the setting is now disabled
* Confirmed that when I visit each agent page and click "Bring this node online", the FreeBSD agent is brought online and is running Java 17
* Confirmed that the Pipeline runs to completion once a FreeBSD agent is online
* Confirmed that the nodeMonitors configuration as code uses the following successfully:
```yaml
  nodeMonitors:
  - diskSpaceMonitor:
      freeSpaceThreshold: "1GB"
  - tmpSpace:
      freeSpaceThreshold: "1GB"
  - jvmVersion:
      comparisonMode: RUNTIME_GREATER_OR_EQUAL_MASTER_BYTECODE
      disconnect: false
  - "remotingVersion"
```

#### 2.444 and current plugin release

* Confirmed that the nodeMonitors configuration as code passes Jenkins startup when using:
```yaml
  nodeMonitors:
  - diskSpaceMonitor:
      freeSpaceThreshold: "1GB"
  - tmpSpace:
      freeSpaceThreshold: "1GB"
  - jvmVersion:
      comparisonMode: RUNTIME_GREATER_OR_EQUAL_MASTER_BYTECODE
      disconnect: false
  - "remotingVersion"
```
* Confirmed that disabling or enabling "Don't mark agents temporarily offline" and clicking "Apply" does not change the connection status of the FreeBSD agent that is running Java 17 and connected to a Java 21 controller

#### 2.440.1-rc and this pull request

* Confirmed that the nodeMonitors configuration as code loads correctly when using:
```yaml
  nodeMonitors:
  - diskSpaceMonitor:
      freeSpaceThreshold: "1GB"
  - tmpSpace:
      freeSpaceThreshold: "1GB"
  - jvmVersion:
      comparisonMode: RUNTIME_GREATER_OR_EQUAL_MASTER_BYTECODE
      disconnect: false
  - "remotingVersion"
```
* Confirmed that the agents are immediately taken offline when the new checkbox "Don't mark agents temporarily offline" is disabled
* Confirmed that agents are not taken offline when the checkbox "Don't mark agents temporarily offline" is enabled
* Confirmed that agents are brought online immediately after the "Don't mark agents temporarily offline" checkbox is disabled and the "Apply" button is clicked

The new behavior of immediately taking the agent offline and immediately bringing it online when the change is applied is very, very nice.  Thanks!

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by viceice** [comment]
Anything missing here? i really like to have the warning automatically disabled when node is online again. Currently this needs a jenkins restart

**Comment by MarkEWaite** [comment]
> Anything missing here? i really like to have the warning automatically disabled when node is online again. Currently this needs a jenkins restart

I was waiting for the pull request to no longer have conflicting files and to not be a draft pull request.  Are you ready for it to be reviewed?

**Comment by mawinter69** [comment]
I need to review the behaviour and remove the now duplicate disconnect for the jvm monitor

**Comment by mawinter69** [comment]
@MarkEWaite fixed the casc issue

File: src/main/java/hudson/plugin/versioncolumn/JVMVersionMonitor.java [hunk 1]
```diff
@@ -46,19 +52,31 @@ public class JVMVersionMonitor extends NodeMonitor {
...
 
     @SuppressWarnings("unused") // jelly
     public boolean isDisconnect() {
-        return disconnect;
+        return !isIgnored();
+    }
+
+    // should be restricted/deprecated but that breaks casc
+    @DataBoundSetter
+    public void setDisconnect(boolean disconnect) {
```
Comment (mawinter69) [inline src/main/java/hudson/plugin/versioncolumn/JVMVersionMonitor.java:71]: That method is required for backward compatibility with CasC. Ideally it should be deprecated (likewise isDisconnected should also be deprecated). But when doing so CasC will fail unless one adds
```
configuration-as-code:
  version: 1
  deprecated: warn
```
to the CasC config.
Similarly using `@Restricted(DoNotUse.class)` makes CasC fail

**Review by MarkEWaite — APPROVED** [review]
Thanks very much.  Testing looks good.  Ready to release.

**Pull request merged**